catchup/service.go

```diff
@@ -464,3 +491,55 @@ func (s *Service) sync() {
 
 	s.log.Infof("Catchup Service: finished catching up, now at round %v (previously %v). Total time catching up %v.", s.ledger.LastRound(), pr, elapsedTime)
 }
+
+// nextRoundIsNotApproved returns true if the next round upgrades to a protocol version
+// which is not approved.
+// In case of an error, it returns false
+func (s *Service) nextRoundIsNotApproved(nextRound basics.Round) bool {
+	lastLedgerRound := s.ledger.LastRound()
+	proto, err := s.ledger.ConsensusParams(lastLedgerRound)
+	if err != nil {
+		s.log.Errorf("catchup: could not get consensus parameters for round %v: $%v", lastLedgerRound, err)
+		s.log.Errorf("catchup: could not determine if next round is from an approved protocol.")
+		return false
+	}
+	approvedUpgrades := proto.ApprovedUpgrades
+
+	block, error := s.ledger.Block(lastLedgerRound)
+	if error != nil {
+		s.log.Errorf("catchup: could not last block (%v) from the ledger.", lastLedgerRound)
+		s.log.Errorf("catchup: could not determine if next round is from an approved protocol.")
+		return false
+	}
+	bh := block.BlockHeader
+	_, isAnApprovedUpgrade := approvedUpgrades[bh.NextProtocol]
+	if bh.NextProtocolSwitchOn > 0 &&
+		nextRound >= bh.NextProtocolSwitchOn &&
+		!isAnApprovedUpgrade {
+		return true
+	}
+	return false
+}
+
+// handleUnapprovedRound records the last approved round number.
+// Checks if the last approved round was added to the ledger, and stops the service.
+func (s *Service) handleUnapprovedRound(nextUnapprovedRound basics.Round) {
```

Comment:
The function is correct, but only if the following is true:
The function expect to receive a round number r which was verified to be unsupported, while the round r-1 was verified to be supported.
We should add this to the comment above the function, as this is a prerequisite for the functional correctness.

Reply:
This part of the code is confusing, and arguably wrong. 
I am moving the statement that sets s.lastApprovedRound to nextRoundIsNotApproved, and get the value directly from NextProtocolSwitchOn.  

> The function expect to receive a round number r which was verified to be unsupported, while the round r-1 was verified to be supported.

This is not true. It could be that round r-1 is not supported:
Hypothetically, when not supporting beginning at round 1, round 1 and round 2 will be fetched in parallel, and will know that round 1 and 2 are not supported after fetching round 1. Hence, only after round 1 is fetched, handleUnapprovedRound may be called on (round 3). In which case, round 2 (r-1) is not supported.